### DISCUSSION TITLE
Can someone please clear my doubts..

### DISCUSSION THREAD

**Discussion by lalit2005** [vercel/next.js#20894] [Help] [answered]
I have been learning Next.JS lately and I really love it.
( I am a beginner in React.JS also ).  
Even after reading 30+ articles and watching YouTube videos on `getStaticProps`, `getInitialProps`, `getServerSideProps` and `getStaticPaths` I am not able to understand any of those terms/functions clearly,  
**So can someone please clear my doubts like I'm 5?**

---
1. What is `getStaticProps`?
    1. It is documented that this function fetches data on "build time", so does this mean that when we run `next build`, it fetches all the data, and when we deploy or run `npm run start`, it uses the previously fetched data to display it.
    2. Can we use `getStaticProps` to fetch data periodically which does not change frequently?
    3. If I use this to fetch data which requires an API key ( like OpenWeatherMap API ), whether my API key will be exposed on the client side, OR whether the webpage gets prerendered on the server side which in turn provides only the HTML to the browser? 
---
2. I did not understand anything about `getInitalProps`. But I guess it does the same work of `getStaticProps`...
    1. If yes, what is the difference between them
---
3. I literally did not understand anything about ☹️☹️ `getStaticPaths`. Why does Next.JS requires the paths to be predefined?
    1. If there are a millon endpoints ( like amazon ) depending upon a database, then what should be done?
    2. What can be done if we cannot define the routes in prior ( when endpoints depend on user's request )?
---
4. I had previously read that `getServerSideProps` and `getStaticProps` do the same thing, but `getServersideProps` fetches data on server side and pre-renders the webpage on every request. Am I right in the previous line?
    1. Can we accomplish the same task with `getStaticProps`? What would be the pros/cons in this case?
---
#### Please answer this one
5. What about serverless functions😧?? Is any of the above methods an alternative to serverless function?
    1. When we can accomplish most of the tasks with serverless functions, why would someone want to use `getStaticProps`, `getInitialProps`, `getServerSideProps` and `getStaticPaths`?
    2. `getServerSideProps` and serverless functions, both run on server side and not client side, So what is the main difference between them??
---

**Comment by michrome** [comment, marked as answer]
A key thing to understand is that you can have the server generate the page that's going to be sent to the browser every time the page is requested (**server-side rendering**) _or_ you can generate the page once, _before_ it is requested (**static generation**). The second approach is faster to serve but the page must be static.

1. `getStaticProps`
    * `getStaticProps` are properties of your page that never change. You're correct: you can get the data once and generate the page once during `build`.
    * No, you can't use `getStaticProps` to fetch data periodically. You can use [Incremental Static Regeneration](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration) to create a very similar effect though.
    * The code running in `getStaticProps` is not exposed to the browser, so it can use API keys etc.
2. `getInitalProps` is best avoided. Instead pick between `getStaticProps` for static generation or `getServerSideProps` for server-side rendering. 
3. `getStaticPaths` is the list of pages you want to statically generate with `getStaticProps`.
    * Don't statically generate all one million pages; rely on server-side rendering.
    * [Dynamic routes](https://nextjs.org/docs/routing/dynamic-routes) are what you use when you don't know the routes in advance.
4. Your understanding is correct. `getStaticProps` for static generation (faster, less dynamic). `getServerSideProps` for server-side rendering (slower, more dynamic).
5. Serverless functions are not used to generate pages. Typically, serverless functions are used to perform an action.

Hope this helps 😀

**Reply by leerob** [reply]
Great answer! Only one thing to clarify. `getServerSideProps` isn't slower _always_, but only on the first request. If you add caching, then subsequent requests would have the same performance as something statically generated, because they're both cached.

But that first hit is very important. That's one of the biggest advantages of SSG. You guarantee the first hit is fast and available.

**Reply by lalit2005** [reply]
Thanks a lot @michrome  and @leerob  for taking the time to answer my questions. It has really helped me.
But I still have some questions..

> The code running in `getStaticProps` is not exposed to the browser, so it can use API keys etc.

Only `getStaticProps` is not exposed to browser or even the other functions ( `getServerSideProps ` and `getStaticPaths` )  are not exposed?

If I were to make a blog which fetches the posts from a headless CMS, If I use `getStaticProps` how will it render the latest post without `building` the website again. So what should I do in this case?

**Reply by michrome** [reply]
> Great answer! Only one thing to clarify. `getServerSideProps` isn't slower _always_, but only on the first request. If you add caching, then subsequent requests would have the same performance as something statically generated, because they're both cached.
> 
> But that first hit is very important. That's one of the biggest advantages of SSG. You guarantee the first hit is fast and available.

Thanks @leerob! You've muddied the water for me slightly: my understanding is that `getServerSideProps` is used to fetch frequently changing data at request time, on each request. Because it's frequently changing, we can't (more accurately, we don't want to) cache it. Is there a subtlety I am missing?

For anyone else reading this, the [Next.js Pre-rendering and Data Fetching page](https://nextjs.org/learn/basics/data-fetching/two-forms) is a great read, with some useful diagrams 😀.

**Reply by michrome** [reply]
> Thanks a lot @michrome and @leerob for taking the time to answer my questions. It has really helped me.

You're welcome @Lalit2005 😄

> But I still have some questions..
> 
> > The code running in `getStaticProps` is not exposed to the browser, so it can use API keys etc.
> 
> Only `getStaticProps` is not exposed to browser or even the other functions ( `getServerSideProps ` and `getStaticPaths` ) are not exposed?

The code in `getStaticProps`, `getServerSideProps` and `getStaticPaths` is not exposed to the browser. You can verify that using [this tool](https://next-code-elimination.vercel.app).

> If I were to make a blog which fetches the posts from a headless CMS, If I use `getStaticProps` how will it render the latest post without `building` the website again. So what should I do in this case?

For this example, you probably have two pages to update:

1. An “index” page that lists and links to your blog post pages. This page should use [Incremental Static Regeneration](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration) to maintain the list.
2. A new “blog post” page. Because this page didn't exist when you first built the site, you can build it later using `getStaticPaths`’ [`fallback: true`](https://nextjs.org/docs/basic-features/data-fetching#fallback-true) key.

_As an alternative to the above, you could rebuild the site. If your site is hosted on Vercel then your CMS can call a [deploy hook](https://vercel.com/docs/more/deploy-hooks) when a new blog post is published. Personally, I prefer the Incremental Static Regeneration and `fallback` (ISR&F) approach to a deploy hook, because ISR&F avoids purging other unchanged pages from Vercel's CDN._

**Reply by lalit2005** [reply]
Thanks a lot again @michrome,
I understood everything very well, but can you shed some more light on `getStaticPaths`?
I am not able to understand this concept very well...☹️

Can you also please tell how to cache the results of `getServerSideProps` as @leerob said.

And can we also cache `getServerSideProps` results??

**Reply by michrome** [reply]
> Thanks a lot again @michrome,
> I understood everything very well, but can you shed some more light on `getStaticPaths`?
> I am not able to understand this concept very well...☹️

Let's imagine you have two paths `/1` and `/2` that you want to statically generate:

```html
<!-- my-site.com/1 -->
<h1>I am page: One</h1>
```

```html
<!-- my-site.com/2 -->
<h1>I am page: Two</h1>
```

* You'd use `getStaticPaths` to return `/1` and `/2` — to tell Next.js **which paths** to statically generate during the build step.
* You'd use `getStaticProps` to return `One` and `Two` — to tell Next.js **what props/data** to use for each of those paths.

**Reply by lalit2005** [reply]
> Let's imagine you have two paths `/1` and `/2` that you want to statically generate:
> 
> ```
> 
> <h1>I am page: One</h1>
> ```
> 
> ```
> 
> <h1>I am page: Two</h1>
> ```
> 
> * You'd use `getStaticPaths` to return `/1` and `/2` — to tell Next.js **which paths** to statically generate during the build step.
> * You'd use `getStaticProps` to return `One` and `Two` — to tell Next.js **what props/data** to use for each of those paths.

But we can do the same by creating 1.js and 2.js pages in pages/ directory.
What is the advantage of using `getStaticPaths` over adding pages directly in pages directory?

**Reply by michrome** [reply]
> But we can do the same by creating 1.js and 2.js pages in pages/ directory.
> What is the advantage of using `getStaticPaths` over adding pages directly in pages directory?

The use case is your [original question 3](https://github.com/vercel/next.js/discussions/20894#discussion-1637504): when your paths are defined from elsewhere, such as a database or CMS.

**Reply by lalit2005** [reply]
> The use case is your [original question 3](https://github.com/vercel/next.js/discussions/20894#discussion-1637504): when your paths are defined from elsewhere, such as a database or CMS.

We can use Dynamic Routing with a database or CMS right?

**Reply by michrome** [reply]
Yes: that's exactly it! `getStaticPaths` lets you define the dynamic routes (if there are any) you want to statically generate. 😀

**Reply by lalit2005** [reply]
> Yes: that's exactly it! `getStaticPaths` lets you define the dynamic routes (if there are any) you want to statically generate. 😀

So should we use `getStaticPaths` in dynamic files such as [id].js or [post].js  ??

Sorry if I am troubling you by asking a lot of silly questions...

**Reply by michrome** [reply]
Not silly questions at all: they are logical and clear.

Yes, that's exactly where you'd use `getStaticPaths`. Sounds to me like you understand it perfectly. 🥳

**Reply by lalit2005** [reply]
Thanks a lot for the help 🙇 

So if we have less number of dynamic routes for a page, we use `getStaticPaths `. 
And if we have many pages, we use `getServerSideProps` with caching.
Am I right?

One last question...
How can we cache the results of `getServerSideProps` as [@leerob](https://github.com/vercel/next.js/discussions/20894#discussioncomment-270595) said before?

**Reply by michrome** [reply]
I suspect Lee was referring to Vercel (as opposed to Next.js). From the [Vercel caching docs](https://vercel.com/docs/edge-network/caching):

> **Static caching is automatic for all deployments.** This means that no changes need to be made to headers. You get the best performance possible out of the box with zero configuration required.
> 
> Responses with HTTP status 200, 404, 301 or 308, to a GET or HEAD request are cached by our CDN.

**Reply by lalit2005** [reply]
> I suspect Lee was referring to Vercel (as opposed to Next.js). From the [Vercel caching docs](https://vercel.com/docs/edge-network/caching):
> 
> > **Static caching is automatic for all deployments.** This means that no changes need to be made to headers. You get the best performance possible out of the box with zero configuration required.
> > Responses with HTTP status 200, 404, 301 or 308, to a GET or HEAD request are cached by our CDN.

Thanks for the link, it was really helpful.




> Thanks a lot for the help 🙇
> 
> So if we have less number of dynamic routes for a page, we use `getStaticPaths `.
> And if we have many pages, we use `getServerSideProps` with caching.

Am I right?

**Reply by michaelschufi** [reply]
> Great answer! Only one thing to clarify. `getServerSideProps` isn't slower _always_, but only on the first request. If you add caching, then subsequent requests would have the same performance as something statically generated, because they're both cached.

@leerob 
How would you go about caching the page? Is Next.js actually automatically caching the page if it sees that the returned props are the same? If not, how can I prevent the unnecessary re-rendering of the page in this case?

**Reply by lalit2005** [reply]
> > Great answer! Only one thing to clarify. `getServerSideProps` isn't slower _always_, but only on the first request. If you add caching, then subsequent requests would have the same performance as something statically generated, because they're both cached.

Finally found out how to cache `getServerSideProps` pages in one of @leerob 's video 😃 
https://youtu.be/eMwTEo6AjDc?t=3050
